test/distributed/org/apache/cassandra/distributed/test/streaming/BoundExceptionTest.java

```diff
@@ -0,0 +1,144 @@
+
+    @Test
+    public void testSingleException() {
+        try {
+            throw new RuntimeException("test exception");
+        } catch (Exception e) {
+            StreamSession.boundStackTrace(e, LIMIT, new HashSet<>());
+            String expectedStackTrace = "java.lang.RuntimeException: test exception\n" +
+                                        "\tat org.apache.cassandra.distributed.test.streaming.BoundExceptionTest.testSingleException(BoundExceptionTest.java:38)\n" +
+                                        "\tat java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)\n";
+            assertEquals(expectedStackTrace,Throwables.getStackTraceAsString(e));
+            assertEquals(e.getStackTrace().length, LIMIT);
+        }
+    }
+
+    @Test
+    public void testNestedException() {
+        try {
+            throw new RuntimeException(new IllegalArgumentException("the disk /foo/var is bad", new IOException("Bad disk somewhere")));
+        } catch (Exception e) {
+            StreamSession.boundStackTrace(e, LIMIT, new HashSet<>());
+            String expectedStackTrace = "java.lang.RuntimeException: java.lang.IllegalArgumentException: the disk /foo/var is bad\n" +
+                                        "\tat org.apache.cassandra.distributed.test.streaming.BoundExceptionTest.testNestedException(BoundExceptionTest.java:52)\n" +
+                                        "\tat java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)\n" +
+                                        "Caused by: java.lang.IllegalArgumentException: the disk /foo/var is bad\n" +
+                                        "\t... " + LIMIT + " more\n" +
+                                        "Caused by: java.io.IOException: Bad disk somewhere\n" +
+                                        "\t... " + LIMIT + " more\n";
+            assertEquals(expectedStackTrace,Throwables.getStackTraceAsString(e));
+            assertEquals(e.getStackTrace().length, LIMIT);
+        }
+    }
+
+    @Test
+    public void testExceptionCycle()
+    {
+        try {
+            throw new Exception("Test exception 1");
+        }
+        catch (Exception e1) {
+            try {
+                throw new RuntimeException("Test exception 2");
+            }
+            catch (Exception e2) {
+                e1.initCause(e2);
+                e2.initCause(e1);
+
+                causeCycle(e1);
+            }
+        }
+    }
+
+    private static void causeCycle(Exception e) {
+        try {
+            throw e;
+        } catch (Exception e1) {
+            try {
+                StreamSession.boundStackTrace(e1, LIMIT, new HashSet<>());
+            }
+            catch (Exception e2) {
+                String expectedStackTrace = "java.lang.IllegalArgumentException: Exception cycle detected";
+                assertTrue(Throwables.getStackTraceAsString(e2).contains(expectedStackTrace));
```

Comment:
think you will like `org.assertj.core.api.Assertions#assertThat(java.lang.Throwable)` and `org.assertj.core.api.Assertions#assertThatThrownBy(org.assertj.core.api.ThrowableAssert.ThrowingCallable)`, makes testing exception state easier